### PR TITLE
Fixes #4391 stop background processes before rollback and update

### LINKED ISSUE #4391: PHP Fatal error:  Uncaught TypeError: Argument 1 passed to WP_Rocket\Dependencies\Minify\JS when update WPR while RUCSS is in progress
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version => yes
- Used the search feature to ensure that the bug hasn’t been reported before 

**Describe the bug**
Fatal error added to the log file when upgrading plugin while having background process in progress

**To Reproduce**
1. Using version 3.9.4, Enable RUCSS
2. On the plugins page, update WPR to 3.9.4.1 or latest (While RUCSS is still in progress)
3. Check the debug.log

**Expected behavior**
No Error in debug.log file

**Screenshots**
[debug.log](https://github.com/wp-media/wp-rocket/files/7243468/debug.log)

**Additional context**
- Error is not affecting using of the dashboard nor the test site
- We need to check while CPCSS / preload is in progress too
- If we switched between branches in the same scenario without reactivate plugin, we will see different log error

**Backlog Grooming (for WP Media dev team use only)**
- [ ] Reproduce the problem
- [ ] Identify the root cause
- [ ] Scope a solution
- [ ] Estimate the effort

### PR DESCRIPTION
**Pull request by mostafa-hisham**
## Description

stop background processes before rollback and update to avoid fatal if the processes still running while the plugin changes

Fixes #4391 

## Type of change

Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)

## Is the solution different from the one proposed during the grooming?

there is no grooming
## How Has This Been Tested?

- [x] we should create an alpha release with this branch so we can test it in both rollback and upgrade

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by iCaspar — COMMENTED** [review]
What's here looks good.
Can we add integration tests to cover this?

**Comment by mostafa-hisham** [comment]
@iCaspar 
I successfully created Integration tests for 
stop CPCSS background Process
stop homepage preload background process

in both 
stop warmup background process
stop sitemap preload background process

i think the problem is I couldn't start them in the integration test 
to check if the upgrade or rollback action will stop them

you can have a look at the classes in the [previous commit](https://github.com/wp-media/wp-rocket/pull/4392/commits/c4d7f4c06b06bbbae7d3dcb2b0a699875285250c) if you want to give it a try

File: tests/Integration/inc/Engine/Preload/Homepage/stopPreload.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		parent::tearDown();
+		$this->restoreWpFilter( 'wp_rocket_upgrade' );
+		$this->restoreWpFilter( 'admin_post_rocket_rollback' );
+	}
+	public function testShouldStopPreloadRollback() {
+		$this->startPreload();
+		do_action('admin_post_rocket_rollback');
+		$is_cron_exists= wp_get_schedule('rocket_preload_cron');
+
+		$this->assertSame(  false , $is_cron_exists );
```
Comment (iCaspar) [inline tests/Integration/inc/Engine/Preload/Homepage/stopPreload.php]:
```suggestion
		$this->assertFalse( $is_cron_exists );
```

File: tests/Integration/inc/Engine/Preload/Homepage/stopPreload.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	public function testShouldStopPreloadUpgrade() {
+		$this->startPreload();
+		do_action('wp_rocket_upgrade','3.9.4.1', '3.10');
+		$is_cron_exists= wp_get_schedule('rocket_preload_cron');
+
+		$this->assertSame(  false , $is_cron_exists );
```
Comment (iCaspar) [inline tests/Integration/inc/Engine/Preload/Homepage/stopPreload.php]:
```suggestion
		$this->assertFalse( $is_cron_exists );
```

**Review by iCaspar — CHANGES_REQUESTED** [review]
Nice work.

**Review by iCaspar — APPROVED** [review]

**Comment by piotrbak** [comment]
@mostafa-hisham 
When rolling back from the branch to 3.9, there's no fatal error, but the RUCSS process can be stuck on warming up with wrong data:
`"warmup_status":{"total":1,"warmed_count":0,"notwarmed_resources":[],"duration":70,"completed":false}`. 
While updating, it got stuck too, but the data was more accurate.

It seems not to be a problem with this branch, as it's only here to prevent fatal error from happening. There's no pattern to reproduce this though.

@Tabrisrp What do you think? Should we somehow make sure that RUCSS process will be continued and more reliable in this PR?

**Comment by remyperona** [comment]
Since we cancel the process, we can't resume it easily after. I'd say it's ok for this PR, and we can follow up on another issue to handle this case.

**Comment by piotrbak** [comment]
@mostafa-hisham @Tabrisrp After one of the tries, my log got blasted with that kind of notices:
```
[05-Oct-2021 15:40:59 UTC] PHP Notice:  Undefined index: external in /var/www/rocketlab.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Optimization/RUCSS/Warmup/ResourceFetcherProcess.php on line 123

[05-Oct-2021 15:40:59 UTC] PHP Notice:  Undefined index: path in /var/www/rocketlab.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Optimization/RUCSS/Warmup/ResourceFetcherProcess.php on line 123

[05-Oct-2021 15:40:59 UTC] PHP Notice:  Undefined index: path in /var/www/rocketlab.ovh/htdocs/wp-content/plugins/wp-rocket/inc/Engine/Optimization/RUCSS/Warmup/ResourceFetcherProcess.php on line 135

```

Do you think we can do something about that?

**Comment by mostafa-hisham** [comment]
@piotrbak @Tabrisrp 
i guarded the variables to avoid warnings

**Review by piotrbak — APPROVED** [review]
Good to go!
[testrail-report-269.pdf](https://github.com/wp-media/wp-rocket/files/7311611/testrail-report-269.pdf)

**Pull request merged**